docs/development/design/flyout-redesign.rst

```diff
@@ -0,0 +1,165 @@
+which has some basic data about the project and the specific build being served.
+This includes data such as the project and version slugs, the build date, and theme.
+
+One of the goals of this project is to give theme authors and project maintainers integration points
+where they can control some aspects of Read the Docs in a repeatable, deterministic way.
+In order to do this, we need to have JavaScript constants and functions that these maintainers can set.
+This proposal suggests creating a new ``readthedocs`` global variable like so:
+
+.. code-block:: javascript
+
+    // This first line helps handle the case where the project/theme's overrides
+    // are executed before the readthedocs global is defined.
+    window.readthedocs = window.readthedocs || {};
+    window.readthedocs.data = JSON.parse(document.getElementById('READTHEDOCS_DATA').innerHTML);
+
+    if (!window.readthedocs.integration_version_selector) {
+        // Set the default flyout menu code
+        // But only if the user hasn't overridden it already
+        window.readthedocs.integration_version_selector = function (project) {
+            // Existing code we use to create the flyout menu
+            // Currently, this is in `core/static-src/core/js/doc-embed/footer.js:injectFooter`
+            // The `project` variable is the result of the v3 project API for the current project
+        };
+    }
+
+When Read the Docs goes to create the flyout menu,
+it will call this new ``readthedocs.integration_version_selector`` method
+which will either be our default flyout menu method or the overridden method from the theme or project.
+
+This gives us a baseline of how to add these kinds of integrations.
+There are other possible integrations that users may want in the future such as:
+
+* Control how RTD's search overrides Sphinx's built-in search.
+  This could be used to remove RTD's search overrides entirely
+  or to just change how they take effect.
+* Control how RTD's built-in project analytics sends data.
+* There could be an override to how the project's data is retrieved from the API.
+  This could allow not even getting project/version/translation data at all to save an API call for the project.
+
+
+Disabling the flyout entirely
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+One nice aspect of this integration is that for projects or themes that want to completely disable
+the version selector, this could be done by having JS like this:
+
+.. code-block:: javascript
+
+    window.readthedocs = window.readthedocs || {};
+    window.readthedocs.integration_version_selector = function () {};
+
+An alternative would be a way for JS projects to define constants that affects how RTD works.
+This could be something like:
+
+.. code-block:: javascript
+
+    window.readthedocs = window.readthedocs || {};
+    window.readthedocs.customizations = {disable_custom_search: true, disable_version_selector: true};
```

Comment:
we already do this, but we haven't document them. I like more the name you are proposing here p:

`READTHEDOCS_DATA.features.docsearch_disabled`

Reply:
I know we do this. I like the idea a lot but I wish it was easier for users to control as well as documented. The way it is now can run into race conditions with us defining these features and users setting them.